# STOR 235 — Lab 1

This lab is an introduction to Python programming. It provides the fundamental skills you will use to complete the remaining labs in the course.

Collaborating on labs is encouraged! You should rarely remain stuck for more than a few minutes on these questions, so ask a classmate or an instructor for help if you're confused about something. Please don't just share answers, though.  Same goes for AI like ChatGPT -- you might ask 'what is the correct syntax for ...' but you should not be asking AI to complete the assignment for you.

#### Submitting your work

On canvas, you will find the assignment and you should upload your completed .ipynb notebook

#### Today's lab

In today's lab, you will learn how to:

1. Navigate Jupyter notebooks (like this one);
2. Write and evaluate some basic *expressions* in Python, the programming language for the course;
3. Understand how to import modules (collections of functions) to do things, as well as write your own functions;
4. Basic vector manipulation.

There are numerous online resources on the Python language, and you are welcome to seek them out if you would like supplementary references. In particular, [this cheat sheet](https://learnxinyminutes.com/python/) covers basic aspects of the Python language and may be a useful reference. Sections 1 through 5 are the most relevant to us. I **do not** recommend trying to read it straight through; that would be tedious and boring. Rather, I suggest that you consult it when you have a specific question about how to do something.

### Every question has 5 points and looks like **Question X.Y.Z**. Please make sure you do all of the tasks in the bolded questions!

# 1. Jupyter notebooks
A Jupyter notebook is a place to write programs and view their results, and also to write text.

## 1.1. Text cells
In a notebook, each rectangle containing text or code is called a *cell*.

Text cells (like this one) can be edited by double-clicking on them. They're written in a simple format called [Markdown](https://learnxinyminutes.com/markdown/) to add formatting and section headings. (Again, I recommend against reading that reference right now. You can probably pick up what you need to know by looking at the text blocks in this notebook.)

After you edit a text cell, click another cell or press `shift` + `return` to confirm any changes. (Try not to delete the instructions of the lab.)

**Question 1.1.1.** This paragraph is in its own text cell.  Try editing it so that this sentence is the last sentence in the paragraph.

## 1.2. Code cells

Other cells contain code in the Python 3 language. Running a code cell will execute all of the code it contains.

To run the code in a code cell, first click on that cell to activate it.  It will be highlighted.  Next, either press ▶| or hold down `shift` + `return`.

Try running this cell:

In [47]:
print("Hello, World!")

Hello, World!


As the name suggests, the `print` command prints out text. You will need to put the text you want to print in quotes, to distinguish it from Python commands.

The `print` command can also print emojis. There are special commands beginning with `\N` to get the emojis. (A full list can be found online.)

In [48]:
print("\N{WAVING HAND SIGN}, \N{EARTH GLOBE AMERICAS}!")

👋, 🌎!


The fundamental building block of Python code is an expression. Cells can contain multiple lines with multiple expressions. When you run a cell, the lines of code are executed in the order in which they appear. Every `print` expression prints a line. Run the next cell and notice the order of the output.

In [49]:
print("First this line,")
print("then the whole \N{EARTH GLOBE AMERICAS},")
print("and then this one.")

First this line,
then the whole 🌎,
and then this one.


**Question 1.2.1.** Change the cell above so that it prints out:

    First this line,
    then the whole 🌎,
    and then this one.

*Hint:* If you're stuck on the Earth symbol for more than a few minutes, try talking to a classmate or a course staff member. That's a good idea for any lab problem.

## 1.3. Writing Jupyter notebooks
You can use Jupyter notebooks for your own projects or documents.  When you make your own notebook, you'll need to create your own cells for text and code.

To add a cell, click the + button in the menu bar. You can use `+ Code` for a code cell, and `+ Text` for a text cell.

**Question 1.3.1.** Add a code cell below this one.  Write code in it that prints out:
   
    A whole new cell! 🎵🌎🎵

(That musical note symbol is like the Earth symbol.  Its long-form name is `\N{MUSICAL NOTE}`.)

Run your cell to verify that it works.

In [50]:
print("A whole new cell! \N{MUSICAL NOTE}\N{EARTH GLOBE AMERICAS}\N{MUSICAL NOTE}")

A whole new cell! 🎵🌎🎵


## 1.4. Errors
Python is a language, and like natural human languages, it has rules.  It differs from natural language in two important ways:
1. The rules are *simple*.  You can learn most of them in a few weeks and gain reasonable proficiency with the language in a semester.
2. The rules are *rigid*.  If you're proficient in a natural language, you can understand a non-proficient speaker, glossing over small mistakes.  A computer running Python code is completely literal in its interpretation.

When you write code, you may make mistakes.  When you run a code cell that has errors, Python will produce error messages to tell you what you did wrong.

Errors are okay. Even experienced programmers make many errors. When you make an error, you just have to find the source of the problem, fix it, and move on.

We have made an error in the next cell.  Run it and see what happens.

In [51]:
print("This line is missing something.")

This line is missing something.


**Note:** In the toolbar, there is the option to click `Run > Run All Cells`, which will run all the code cells in this notebook in order. However, the notebook stops running code cells if it hits an error, like the one in the cell above.

The last line of the error output attempts to tell you what went wrong.  The *syntax* of a language is its structure, and this `SyntaxError` tells you that you have created an illegal structure.  "`EOF`" means "end of file," so the message is saying Python expected you to write something more (in this case, a right parenthesis) before finishing the cell.

There's a lot of terminology in programming languages, but you don't need to know it all in order to program effectively. If you see a cryptic message like this, you can often get by without deciphering it.  (Of course, if you're frustrated, ask a classmate or a course staff member for help.)

**Question 1.4.1.** Fix the code above so that you can run the cell and see the intended message instead of an error.

## 1.5. The Kernel
The kernel is a program that executes the code inside your notebook and outputs the results.

Next to every code cell, you'll see some text that says `In [...]`. Before you run the cell, you'll see `In [ ]`. When the cell is running, you'll see `In [*]`. If you see an asterisk (\*) next to a cell that doesn't go away, it's likely that the code inside the cell is taking too long to run, and it might be a good time to interrupt the kernel (discussed below). When a cell is finished running, you'll see a number inside the brackets, like so: `In [1]`. The number corresponds to the order in which you run the cells; so, the first cell you run will show a 1 when it's finished running, the second will show a 2, and so on.

You may run into problems where your kernel is stuck for an excessive amount of time, your notebook is very slow and unresponsive, or your kernel loses its connection. If this happens, try the following steps:
1. At the top of your screen, click **Kernel**, then **Interrupt**.
2. If that doesn't help, click **Kernel**, then **Restart**. If you do this, you will have to run your code cells from the start of your notebook up until where you paused your work.

# 2. Numbers

Quantitative information arises everywhere in data science. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number 3.25. (Run the cell and see.)

In [52]:
3.2500

3.25

Notice that we didn't have to `print`. When you run a notebook cell, if the *last* line has a numerical value, then Jupyter helpfully prints out that value for you. However, it won't print out prior lines automatically.

In [53]:
print(2)
3
4

2


4

Above, you should see that 4 is the value of the last expression, 2 is printed, but 3 is not because it was neither printed nor last.

## 2.1. Arithmetic
The line in the next cell subtracts.  Its value is what you'd expect.  Please run it.

In [54]:
3.25 - 1.5

1.75

Many basic arithmetic operations are built into Python.

An important common operator that differs from typical mathematical notation is `**`, which raises one number to the power of the other. So, `2**3` stands for $2^3$ and evaluates to 8.

The order of operations is the same as what you learned in elementary school, and Python also has parentheses.  For example, compare the outputs of the cells below.

In [55]:
2 + 3**2

11

In [56]:
(2+3)**2

25

**Question 2.1.1.** Write a Python expression in this next cell that's equal to $11 \times  2^{0.5} - \frac{7}{33} + 3$.

Replace the ellipses (`...`) with your expression.

In [57]:
11 * 2**0.5 - 7/33 + 3

18.344227973982832

# 3. Names
In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we just say, "Bear!"

In Python, we do this with *assignment statements*. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right.

In [58]:
my_number = 3 * 2 + 4

When you run that cell, Python first computes the value of the expression on the right-hand side, `3 * 2 + 4`, which is the number 10.  Then it assigns that value to the name `my_number`.  At that point, the code in the cell is done running.

After you run that cell, the value 10 is bound to the name `my_number`:

In [59]:
my_number

10

The statement `my_number = 3 * 2 + 4` is not asserting that `my_number` is already equal to `3 * 2 + 4`, as we might expect by analogy with mathematical notation.  Rather, that line of code changes what `my_number` means; it now refers to the value 10, whereas before it meant nothing at all.

If the designers of Python had been very pedantic, they might have made us write

    define the name my_number to hereafter have the value of 3 * 2 + 4

instead.  You will probably appreciate the brevity of "`=`"!  But keep in mind that this is the real meaning.

**Question 3.0.1.** Try writing code that uses a name (like `applesauce`) that hasn't been assigned to anything.  You'll see an error! Leave this error in your submitted notebook.

In [ ]:
applesauce

A common pattern in Jupyter notebooks is to assign a value to a name and then immediately evaluate the name in the last line in the cell so that the value is displayed as output.

In [61]:
close_to_pi = 355/113
close_to_pi

3.1415929203539825

## 3.1. Computations with Names

Another common pattern is that a series of lines in a single cell will build up a complex computation in stages, naming the intermediate results.

In [62]:
semimonthly_salary = 2205.78
monthly_salary = 2 * semimonthly_salary
number_of_months_in_a_year = 12
yearly_salary = number_of_months_in_a_year * monthly_salary
yearly_salary

52938.72

Names in Python can have letters (upper- and lower-case letters are both okay and count as different letters), underscores, and numbers.  The first character can't be a number (otherwise a name might look like a number).  And names can't contain spaces, since spaces are used to separate pieces of code from each other.

Other than those rules, what you name something doesn't matter *to Python*.  For example, this cell does the same thing as the above cell, except everything has a different name:

In [63]:
a = 2205.78
b = 2 * a
c = 12
d = c * b
d

52938.72

However, names are very important for making your code *readable* to yourself and others.  The cell above is shorter, but it's totally useless without an explanation of what it does.

**Question 3.1.1.** Assign the name `minutes_in_a_decade` to the number of minutes between midnight January 1, 2013 and midnight January 1, 2023. Note that there are two leap years in this span of a decade (in 2016 and 2020). A non-leap year has 365 days and a leap year has 366 days. It may be easiest if you build up to the answer using a series of auxiliary computations and variables.

In [64]:
# Change the next line
# so that it computes the number of minutes in a decade
# and assigns that number the name minutes_in_a_decade.

regular_years = 8
leap_years = 2

total_days = (regular_years * 365) + (leap_years * 366)
hours_per_day = 24
minutes_per_hour = 60

minutes_in_a_decade = total_days * hours_per_day * minutes_per_hour

# We've put this line in this cell
# so that it will print the value you've given to minutes_in_a_decade when you run it.
# You don't need to change this.
minutes_in_a_decade

5258880

# 3.2. Comments
You may have noticed these lines in the cell in which you answered Question 3.1.1:

    # Change the next line
    # so that it computes the number of minutes in a decade
    # and assigns that number the name, minutes_in_a_decade.
    
This is called a *comment*. It doesn't make anything happen in Python; Python ignores anything on a line after a `#`.  Instead, it's there to communicate something about the code to you, the human reader. Comments are extremely useful.

# 4 Importing modules

This section explains the role and purpose of modules. We will mainly focus on modules relevant for our course, such as `math` and `numpy`.



# 4.1 Importing code

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to import modules. A module is a file with Python code that has defined variables and functions. By importing a module, we are able to use its code in our own notebook.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python.

Suppose we want to very accurately compute the area of a circle with a radius of 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [65]:
# Just run this cell

import math
radius = 5
area_of_circle = radius**2 * math.pi
area_of_circle

78.53981633974483

In the code above, the line `import math` imports the math module. We are now able to access any variables or functions defined within `math` by typing the name of the module followed by a dot, then followed by the name of the variable or function we want.

    <module name>.<name>

**Question 4.1.1** The module `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71. Compute $\sqrt{e^{2\pi}+30\pi}$, giving it the name `near_twentyfive`.

*Remember: You can access `pi` from the `math` module as well!*

<!--
BEGIN QUESTION
name: q41
-->

In [78]:
import math
near_twentyfive = math.sqrt(math.e ** (2*math.pi) + 30*math.pi)
near_twentyfive

25.094609682807544

## 4.2. Accessing functions in packages

In the question above, you accessed variables within the `math` module.

Modules also define functions.  For example, `math` provides the name `log` for the natural logarithm function. Having imported `math` already, we can write `math.log(3)` to compute the natural log of 3.

For your reference, below are some more examples of functions from the `math` module.

Notice how different functions take in different numbers of arguments. Often, the [documentation](https://docs.python.org/3/library/math.html) of the module will provide information on how many arguments are required for each function.

In [67]:
# Calculating logarithms (the logarithm of 8 in base 2).
# The result is 3 because 2 to the power of 3 is 8.
math.log(8, 2)

3.0

In [68]:
# Calculating square roots.
math.sqrt(5)

2.23606797749979

There are various ways to import and access code from outside sources. The method we used above — import `<module_name>` — imports the entire module and requires that we use `<module_name>.<name>` to access its code.
We can also import a specific constant or function instead of the entire module. Notice that you don't have to use the module name beforehand to reference that particular value. However, you do have to be careful about reassigning the names of the constants or functions to other values!

In [69]:
# Importing just cos and pi from math.
# We don't have to use `math.` in front of cos or pi
from math import cos, pi
print(cos(pi))

# We do have to use it in front of other functions from math, though
math.log(pi)

-1.0


1.1447298858494002

Or we can import every function and value from the entire module.

In [70]:
# Lastly, we can import everything from math using the *
# Once again, we don't have to use 'math.' beforehand
from math import *
log(pi)

1.1447298858494002

Don't worry too much about which type of import to use. It is a matter of personal preference. In this course, you'll always import the necessary modules when you run the setup cell (like the first code cell in this lab).

# 4.3 Creating your own functions

In addition to using functions from packages, creating your own functions will be of fundamental importance. The following examples demonstrate the basic syntax.

In [71]:
def my_function(c):
  output = c*9/5 + 32
  print(output)

In [72]:
my_function(100)

212.0


In [73]:
def my_sum(a,b):
  print(a+b)

my_sum(3,5)

8


**Question 4.3.1** Write two functions (call these arithmetic_mean and geometric_mean) each of which accept three numbers (say $a,b,c$) as arguments and:

-  arithmetic_mean: outputs $\frac{a+b+c}{3}$.
-  geometric_mean: outputs $(abc)^{1/3}$.

In [79]:
# Arithmetic average
def arithmetic_mean(a, b, c):
    top = a + b + c
    bottom = 3
    print(top/bottom)

In [80]:
# Geometric average
def geometric_mean(a, b, c):
    print(a*b*c/3)

# 5 Basic vector operations

 Now we will discuss basic ways to define and manipulate vectors. The main package we will need is called `numpy`.



In [82]:
import numpy as np

For creating vectors, we use `np.array`. The `len` command (for "length") returns the dimension of the vector. (The `np.array` command does not distinguish between column and row vectors.)

In [83]:
x = np.array([1,2, 3, 4])
print(x)

[1 2 3 4]


In [84]:
len(x)

4

Specific elements can be extraced using x[i] where the index runs from 0 to the length of the vector minus one. (This is called zero-based indexing, and is different from the one-based indexing that we use in class.)

In [85]:
print(x[0], x[3])

1 4


One common mistake for obtaining the "last" element is using x[4]. See what happens when you do this.

In [86]:
print(x[4])

IndexError: index 4 is out of bounds for axis 0 with size 4

One can use the above to change the index of a specific entry.

In [87]:
x[2]= 20000
print(x)

[    1     2 20000     4]


Vector addition and scalar multiplication is done in the usual manner:

In [88]:
x= np.array([1,2,3])
y= np.array([-1,-2,-3])
print(x+y)
print(2*x)

[0 0 0]
[2 4 6]


**Question 5.1.1** Write code to make a function (called `fun_sum`) that takes as its four arguments two vectors $\mathbf{x}, \mathbf{y}$ and two scalars $\alpha, \beta$, and outputs $\alpha \mathbf{x} + \beta \mathbf{y}$.

In [89]:
#def fun_sum(u,v,a,b):
def fun_sum(x, y, alpha, beta):
    print(alpha*x + beta*y)

Run your function on $\mathbf{x}$ and $\mathbf{y}$ defined above (`x= np.array([1,2,3])` and `
y= np.array([-1,-2,-3])`) and with $\alpha =1$, $\beta =5$, and make sure your result appears as output.

In [90]:
## Your result
fun_sum(x, y, 1, 5)

[ -4  -8 -12]


There is also a function for the inner product (dot product) of two vectors.

In [91]:
a = np.inner(x,y)
print(a)

-14


## Matrices:

We will talk about matrices in a few lectures. For now, just think of them as arrays of numbers. The following cells go over basic matrix operations.

The following code creates and prints the matrix

$$\begin{equation}
B= \left[ \begin{array}{rrr} 1 & 2 & 3 \\ 3 & 2 & 1 \end{array}\right]
\end{equation}.$$


In [95]:
B = np.array([[1, 2, 3],[3, 2, 1]])

B

array([[1, 2, 3],
       [3, 2, 1]])

Operations on matrices typically have the same syntax as vector operations. However, there is one important exception. Matrix multiplicaton, which we will cover later, uses the `@` symbol.

In [96]:
import numpy as np
A = np.array([[1, 3],[2,1]])
B = np.array([[3, 0, 4],[-1, -2, 1]])
C = np.array([[-2, 1],[4, 1]])
D = np.array([[2],[6]])
print(D)

[[2]
 [6]]


In [97]:
print(3*A,'\n')
print(A+C,'\n')
print(A@B)

[[3 9]
 [6 3]] 

[[-1  4]
 [ 6  2]] 

[[ 0 -6  7]
 [ 5 -2  9]]


**Question 5.2.1** Create a 2 by 2 matrix containing the entries 1, 2, 3, 4, and print it. The order of the entries doesn't matter.

In [102]:
# Your solution here
C = np.array([[1, 2], [3, 4]])
print(C)

[[1 2]
 [3 4]]


# **This is the end of this Lab!**

